src/server/coprocessor/executor/aggregation.rs

```diff
@@ -0,0 +1,266 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// FIXME: remove later
+#![allow(dead_code)]
+
+use std::rc::Rc;
+
+use tipb::schema::ColumnInfo;
+use tipb::executor::Aggregation;
+use tipb::expression::Expr;
+
+use util::collections::{HashMap, HashMapEntry as Entry};
+use util::codec::datum::{self, DatumEncoder, approximate_size};
+use util::codec::table::RowColsDict;
+use util::xeval::{Evaluator, EvalContext};
+
+use super::super::Result;
+use super::{Executor, Row, ExprColumnRefVisitor};
+use super::super::endpoint::{inflate_with_col, SINGLE_GROUP};
+use super::super::aggregate::{self, AggrFunc};
+
+struct AggregationExecutor<'a> {
+    group_by: Vec<Expr>,
+    aggr_func: Vec<Expr>,
+    group_keys: Vec<Rc<Vec<u8>>>,
+    group_key_aggrs: HashMap<Rc<Vec<u8>>, Vec<Box<AggrFunc>>>,
+    cursor: usize,
+    executed: bool,
+    ctx: EvalContext,
+    cols: Vec<ColumnInfo>,
+    src: &'a mut Executor,
+}
+
+impl<'a> AggregationExecutor<'a> {
+    fn new(mut meta: Aggregation,
+           ctx: EvalContext,
+           columns: &[ColumnInfo],
+           src: &'a mut Executor)
+           -> Result<AggregationExecutor<'a>> {
+        // collect all cols used in aggregation
+        let mut visitor = ExprColumnRefVisitor::new();
+        let group_by = meta.take_group_by().into_vec();
+        try!(visitor.batch_visit(&group_by));
+        let aggr_func = meta.take_agg_func().into_vec();
+        try!(visitor.batch_visit(&aggr_func));
+        // filter from all cols
+        let cols = columns.iter()
+            .filter(|col| visitor.col_ids.contains(&col.get_column_id()))
+            .cloned()
+            .collect();
+
+        Ok(AggregationExecutor {
+            group_by: group_by,
+            aggr_func: aggr_func,
+            group_keys: vec![],
+            group_key_aggrs: map![],
+            cursor: 0,
+            executed: false,
+            ctx: ctx,
+            cols: cols,
+            src: src,
+        })
+    }
+
+    fn get_group_key(&mut self, eval: &mut Evaluator) -> Result<Vec<u8>> {
+        if self.group_by.is_empty() {
+            return Ok(SINGLE_GROUP.to_vec());
+        }
+        let mut vals = Vec::with_capacity(self.group_by.len());
+        for expr in &self.group_by {
+            let v = box_try!(eval.eval(&self.ctx, expr));
+            vals.push(v);
+        }
+        let res = box_try!(datum::encode_value(&vals));
+        Ok(res)
+    }
+
+    fn aggregate(&mut self) -> Result<()> {
+        while let Some(row) = try!(self.src.next()) {
+            let mut eval = Evaluator::default();
+            try!(inflate_with_col(&mut eval, &self.ctx, &row.data, &self.cols, row.handle));
+            let group_key = Rc::new(try!(self.get_group_key(&mut eval)));
+            match self.group_key_aggrs.entry(group_key.clone()) {
```

Comment:
no need to clone here, you can use https://doc.rust-lang.org/std/collections/hash_map/struct.OccupiedEntry.html#method.key later.